### PR TITLE
RFC: Voltage divider sensor

### PR DESCRIPTION
**Pull request by aurel32**
There is already a `voltage-divider` binding defined, which is only used by the `boards/nrf/battery` sample. Looking at the recently added `stm32_vbat` sensor driver, I realized it might be a good idea to implement a more generic sensor using the existing binding, that can replace the `stm32_vbat` sensor and big part of the battery sample. As a bonus it allows to easily measure other voltages with only 2 resistors. It is actually more versatile than existing sample and drivers as it allows to tune all ADC parameters (resolution, oversampling, gain, ...) at the expense of more entries in the device tree.

For now this is just a RFC to demonstrate the idea. I have fully tested the driver itself. I have converted the `sensor/stm32_vbat_sensor` sample to show the changes are minimal, but only tested compiled it, as I don't have the corresponding boards. If this driver sounds like a good idea, I can also convert the `boards/nrf/battery` sample.

### CODE REVIEW

**Comment by carlescufi** [comment]
API meeting: 

- @MaureenHelm @gmarull @erwango

**Comment by aurel32** [comment]
I have just pushed an updated version of this PR, following the recent changes in Zephyr. I have been able to test the `stm32_vbat_sensor` on a Nucleo L432KC board, which have the VBAT pin internally connected to VCC, and the value was correct.

**Comment by aurel32** [comment]
Any news about that? It might be worth looking at the idea before the 3.2 is released with `stm32_vbat`, as this alternative sensor can just replace it and do much more.

**Review by erwango — APPROVED** [review]
I think this addition is valuable and I'm fine if this could replace `stm32_vbat` driver.
Regarding the sample itself, it could be made generic of course but I'd prefer we can keep it identified as `vbat_sensor`, rather than a generic `voltage_divider` sensor, otherwise we'd need some doc addition to inform and redirect people that are looking for a solution to measure vbat.

Edit: Tested ok on supported boards.

File: drivers/sensor/voltage_divider/Kconfig [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# Voltage divider sensor configuration options
+
+# Copyright (c) 2022 Aurelien Jarno
+# SPDX-License-Identifier: Apache-2.0
+
+config VOLTAGE_DIVIDER
+	bool "Voltage divider"
+	default y
+	depends on DT_HAS_VOLTAGE_DIVIDER_ENABLED
+	depends on ADC
```
Comment (gmarull) [inline drivers/sensor/voltage_divider/Kconfig:10]:
```suggestion
	select ADC
```

File: drivers/sensor/voltage_divider/Kconfig [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+	help
+	  Enable driver for Voltage Divider and then also ADC
```
Comment (gmarull) [inline drivers/sensor/voltage_divider/Kconfig:12]: help should be improved, maybe:

```suggestion
	  Enable driver for voltage divider. This driver allows measuring analog voltages scaled with a voltage divider circuit.
```

File: drivers/sensor/voltage_divider/voltage_divider.c [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(voltage_divider, CONFIG_SENSOR_LOG_LEVEL);
+
+#if DT_HAS_COMPAT_STATUS_OKAY(voltage_divider)
+#define DT_DRV_COMPAT voltage_divider
+#else
+#error "No compatible devicetree node found"
+#endif
```
Comment (gmarull) [inline drivers/sensor/voltage_divider/voltage_divider.c:19]:
```suggestion
```

this is mostly redundant with the `DT_HAS_` helpers

File: drivers/sensor/voltage_divider/voltage_divider.c [hunk 4]
```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) 2022 Aurelien Jarno
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/device.h>
```
Comment (gmarull) [inline drivers/sensor/voltage_divider/voltage_divider.c:7]:
```suggestion
#define DT_DRV_COMPAT voltage_divider

#include <zephyr/device.h>
```

File: drivers/sensor/voltage_divider/voltage_divider.c [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+	if (data->power_gpios.port) {
+		rc = gpio_pin_set_dt(&data->power_gpios, 1);
+		if (rc) {
+			LOG_ERR("Failed to set GPIO %s.%u: %d",
+				data->power_gpios.port->name, data->power_gpios.pin, rc);
+		}
+	}
+
+	return rc;
+}
```
Comment (gmarull) [inline drivers/sensor/voltage_divider/voltage_divider.c:53]: I think we should be using the regulator framework here, and `-supply` properties instead.

File: drivers/sensor/voltage_divider/voltage_divider.c [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+	int rc;
+
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_VOLTAGE) {
+		return -ENOTSUP;
+	}
+
+	k_mutex_lock(&data->mutex, K_FOREVER);
+
+	rc = voltage_divider_enable(dev, 1);
+	if (rc) {
```
Comment (gmarull) [inline drivers/sensor/voltage_divider/voltage_divider.c:68]:
```suggestion
	if (rc < 0) {
```
nit: use expressions evaluating to boolean, same for others

File: drivers/sensor/voltage_divider/voltage_divider.c [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+	int32_t mv;
+	float voltage;
+
+	if (chan != SENSOR_CHAN_VOLTAGE) {
+		return -ENOTSUP;
+	}
+
+	/* Sensor value in millivolts */
+	mv = data->raw;
+	adc_raw_to_millivolts_dt(&data->adc_channel, &mv);
```
Comment (gmarull) [inline drivers/sensor/voltage_divider/voltage_divider.c:106]: this needs error checking (`-ENOTSUP`), not all drivers support this operation.

File: drivers/sensor/voltage_divider/voltage_divider.c [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+	static struct voltage_divider_data voltage_divider_dev_data_##inst = { \
+		.adc_channel = ADC_DT_SPEC_INST_GET(inst), \
+		.power_gpios = GPIO_DT_SPEC_INST_GET_OR(inst, power_gpios, {0}) \
+	}; \
+	DEVICE_DT_INST_DEFINE(inst, voltage_divider_init, NULL, \
+		&voltage_divider_dev_data_##inst, &voltage_divider_dev_config_##inst, \
+		POST_KERNEL, \
+		CONFIG_SENSOR_INIT_PRIORITY, &voltage_divider_driver_api);
+
+DT_INST_FOREACH_STATUS_OKAY(DEFINE_VOLTAGE_DIVIDER)
```
Comment (gmarull) [inline drivers/sensor/voltage_divider/voltage_divider.c:174]: please align `\`, or try with `clang-format`.

File: samples/sensor/stm32_vbat_sensor/README.rst [hunk 9]
```diff
@@ -7,7 +7,7 @@ Overview
 ********
 
 This sample reads the Vbat from the STM32 Internal
```
Comment (gmarull) [inline samples/sensor/stm32_vbat_sensor/README.rst:9]: we need to rename the sample to not be stm32 specific, this is now a generic sample that can work on any board, and can measure any voltage, not just battery.

**Review by gmarull — CHANGES_REQUESTED** [review]

**Comment by erwango** [comment]
@MaureenHelm Should CODEOWNERS/MAINTAINERS be updated ?

File: samples/sensor/stm32_vbat_sensor/README.rst [hunk 10]
```diff
@@ -19,20 +19,30 @@ board DT file or with a board overlay in the samples/sensor/stm32_temp_sensor/bo
 .. code-block:: dts
 
     stm32_vbat: stm32vbat {
-        compatible = "st,stm32-vbat";
+        compatible = "voltage_divider";
```
Comment (FRASTM) [inline samples/sensor/stm32_vbat_sensor/README.rst:22]: compatible = "voltage**-**divider";

File: drivers/sensor/voltage_divider/voltage_divider.c [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+	/* output_ohm is used as a flag value: if it is nonzero then
+	 * the voltage is measured through a voltage divider;
+	 * otherwise it is assumed to be directly connected.
+	 */
+	uint32_t output_ohms;
+	uint32_t full_ohms;
+};
+
+static int voltage_divider_enable(const struct device *dev, bool enable)
+{
```
Comment (FRASTM) [inline drivers/sensor/voltage_divider/voltage_divider.c:40]: what is the difference when enable is true/false -> should 'enable' or disable but there is no check on 'enable' boolean

**Review by erwango — COMMENTED** [review]
Discard comment

File: samples/sensor/stm32_vbat_sensor/boards/nucleo_wb55rg.overlay [hunk 12]
```diff
@@ -3,19 +3,33 @@
...
 };
 
 &adc1 {
 	has-vbat-channel;
 	vref-mv = <3300>;
+
+	#address-cells = <1>;
+	#size-cells = <0>;
+
+	channel@12 {
```
Comment (erwango) [inline samples/sensor/stm32_vbat_sensor/boards/nucleo_wb55rg.overlay:28]:
```suggestion
	channel@18 {
```
?

**Review by MaureenHelm — COMMENTED** [review]
I like the idea and agree with most of the feedback from @gmarull 

> @MaureenHelm Should CODEOWNERS/MAINTAINERS be updated ?

@erwango how?

**Comment by erwango** [comment]
> @erwango how?

From recent examples, I think that CODEOWNER is still in use. So if confirmed, this file should be updated.

**Comment by talih0** [comment]
@aurel32, are you still working on this PR? I can take over if you are too busy (I have some interest in getting this merged upstream)

**Pull request closed without merging**